clients/src/main/java/org/apache/kafka/common/internals/SecurityManagerCompatibility.java

```diff
@@ -0,0 +1,355 @@
+ */
+public interface SecurityManagerCompatibility {
+
+    static SecurityManagerCompatibility get() {
+        return Composite.INSTANCE;
+    }
+
+    <T> T doPrivileged(PrivilegedAction<T> action);
+
+    Subject current();
+
+    <T> T callAs(Subject subject, Callable<T> action) throws CompletionException;
+
+    interface Loader {
+        Class<?> loadClass(String className) throws ClassNotFoundException;
+
+        static Loader forName() {
+            return className -> Class.forName(className, true, Loader.class.getClassLoader());
+        }
+    }
+
+    /**
+     * This class implements reflective access to the deprecated-for-removal methods of AccessController and Subject.
+     * <p>Instantiating this class may fail if any of the required classes or methods are not found.
+     * Method invocations for this class may fail with {@link UnsupportedOperationException} if all methods are found,
+     * but the operation is not permitted to be invoked.
+     * <p>This class is expected to be instantiable in JRE >=8 until the removal finally takes place.
+     */
+    @SuppressWarnings("unchecked")
+    class Legacy implements SecurityManagerCompatibility {
+
+        private final Method doPrivileged;
+        private final Method getContext;
+        private final Method getSubject;
+        private final Method doAs;
+
+        // Visible for testing
+        Legacy(Loader loader) throws ClassNotFoundException, NoSuchMethodException {
+            Class<?> accessController = loader.loadClass("java.security.AccessController");
+            doPrivileged = accessController.getDeclaredMethod("doPrivileged", PrivilegedAction.class);
+            getContext = accessController.getDeclaredMethod("getContext");
+            Class<?> accessControlContext = loader.loadClass("java.security.AccessControlContext");
+            Class<?> subject = loader.loadClass(Subject.class.getName());
+            getSubject = subject.getDeclaredMethod("getSubject", accessControlContext);
+            // Note that we use the real Subject.class rather than the `subject` variable
+            // This allows for mocking out the method without changing the argument types.
+            doAs = subject.getDeclaredMethod("doAs", Subject.class, PrivilegedExceptionAction.class);
+        }
+
+        @Override
+        public <T> T doPrivileged(PrivilegedAction<T> action) {
+            try {
+                return (T) doPrivileged.invoke(null, action);
+            } catch (IllegalAccessException e) {
+                throw new UnsupportedOperationException(e);
+            } catch (InvocationTargetException e) {
+                Throwable cause = e.getCause();
+                if (cause instanceof RuntimeException) {
+                    throw (RuntimeException) cause;
+                } else {
+                    throw new RuntimeException(cause);
+                }
+            }
+        }
```

Comment:
This same stanza is repeated several times, with the only difference being a special clause for `PrivilegedActionException` in `doAs`. Is there a chance we could deduplicate?

Edit: It's also used for other reflective operations in other classes, would be nice to see those covered too. Bare minimum we could have something like a `RuntimeException maybeWrap(Throwable cause)` utility method.

Reply:
I'll deduplicate the try-catch in these new classes, for some reason I thought that it would be more clear if it was verbose. I like the deduplicated form much better.

I won't implement a `maybeWrap(Throwable)` as it has a much larger scope and the benefit is much more unclear to me.